homeassistant/components/faadelays/__init__.py

```diff
@@ -0,0 +1,85 @@
+"""The FAA Delays integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+from aiohttp import ClientConnectionError
+from async_timeout import timeout
+from faadelays import Airport, get_airport_delays
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ID
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["binary_sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the FAA Delays component."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up FAA Delays from a config entry."""
+    code = entry.data[CONF_ID]
+
+    coordinator = FAADataUpdateCoordinator(hass, code)
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class FAADataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching FAA API data from a single endpoint."""
+
+    def __init__(self, hass, code):
+        """Initialize the coordinator."""
+        self.session = aiohttp_client.async_get_clientsession(hass)
+        self.data = Airport(code, self.session)
+        self.code = code
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=timedelta(minutes=1)
+        )
+
+    async def _async_update_data(self):
+        try:
+            with timeout(10):
+                self.data = await get_airport_delays(self.code, self.session)
```

Comment:
Why are we overwriting the `Airport` instance here? Does `get_airport_delays` also return an `Airport` instance?

Reply:
It does, yes. The alternative would be to call the `update` function on the existing object, and I'm not sure why I didn't do that to begin with as it would have the exact same result. I can rewrite this fairly easily to just call the `update` function if that would be better, not sure if that would make any difference. 

I think (would need to test it) that the new line could just be `self.data.update()`.

Reply:
Yes, please do that. It makes more sense to update the existing instance.

Reply:
Tried that and it didn't work, I'm working on figuring out why. It seems that if I call `self.data.update()` or try and pull anything out of the `self.data` object (which should have already been created as an `Airport` instance), it gives me a `NoneType does not have function`... error as if the object has not yet been created. I'm a bit stumped on that to be honest, it's happening when the entity is being created in the `async_setup_entry` function when it calls an `async_refresh()`. As far as I can tell, the `Airport` object should already exist in `self.data` but for some reason it does not. Am I missing something here?

Reply:
Please push what you changed so we can have a look together.

Reply:
The latest commit has the change and is broken.

Reply:
Aha, I found it. I called the `super.__init__()` after setting the `self.data` and the `super.__init__()` was resetting `self.data` to `None`. Fix incoming.